junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/TestFactoryTestDescriptor.java

```diff
@@ -131,11 +140,8 @@ protected void invokeTestMethod(JupiterEngineExecutionContext context, DynamicTe
 		return Optional.empty();
 	}
 
-	private JUnitException invalidReturnTypeException(Throwable cause) {
-		String message = String.format(
-			"@TestFactory method [%s] must return a Stream, Collection, Iterable, or Iterator of %s.",
-			getTestMethod().toGenericString(), DynamicNode.class.getName());
-		return new JUnitException(message, cause);
+	static Optional<TestSource> computeOptionalTestSource(DynamicNode node) {
```

Comment:
Please rename to `getCustomTestSource()` or similar.

Reply:
Replaced the method by an inline assignment.

Reply:
👍